### PR TITLE
#12 bug fix: Split userIDs and call the user info API

### LINKED ISSUE #12: No reply when a user calls Auriga specifying a reaction which reacted many people.
## Overview

- Auriga can not reply message when it is called by a user who specifies a reaction that is reacted by many people (more than 31: I confirmed this number) due to an Error coming back from Slack API.

## Expectation

- Auriga can also reply to this case.

## Error Logs

```
Failed to reply error: failed to get user info: too_many_users
```
`too_many_users` is an error message that Slack API can return: https://api.slack.com/methods/users.info#errors

### PR DESCRIPTION
**Pull request by sho0126hiro**
close #12 
## Overview

- #12 
- split userIDs (API params) before Auriga calls the userInfoAPI.

## Changes

- add `SplitStringSliceInChunks` function in `app/pkg/slice` to enable to create chunked slices from a slice and chunkSize.
- add `chunkedListUsersEmail` method in `slackReactionUsersService` that is able to split a slice (userIDs) before calling user info API, and it can call the API for each chunked slice. 
- I decide `ChunkSizeOfChunkedListUserEmail = 20` because I confirm that this chunkSize shoud be set less than 30, but its a number we can afford.

## Scope of Impact

- No impact on users but it affects how Auriga calls userInfoAPI.

## How to check the operation

I've checked Auriga's operation on a test environment.

### CODE REVIEW

**Review by ichiki-mfw — APPROVED** [review]
LGTM！感謝！

File: app/pkg/slice/string.go [hunk 1]
```diff
@@ -29,3 +29,16 @@ func ToStringSet(s []string) []string {
 	}
 	return set
 }
+
+// SplitStringSliceInChunks SplitStringSliceInChunk split slice in chunk
```
Comment (hotakahattori) [inline app/pkg/slice/string.go:33]: 1 未満の値を指定した場合の動作を書いたほうが良いと思う

File: app/pkg/slice/string.go [hunk 2]
```diff
@@ -29,3 +29,16 @@ func ToStringSet(s []string) []string {
...
+// TODO: Update 1.18+ and use generics
+func SplitStringSliceInChunks(s []string, chunkSize int) (chunks [][]string) {
+	if chunkSize < 1 {
+		return append(chunks, s)
+	}
+	for chunkSize < len(s) {
+		chunks = append(chunks, s[:chunkSize])
```
Comment (hotakahattori) [inline app/pkg/slice/string.go:40]: make で cap か length 指定できないかしら (len(s) を chunkSize で割った商切り上げれば良いと思う)

File: app/pkg/slice/string_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+package slice
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestSplitStringSliceInChunks(t *testing.T) {
```
Comment (hotakahattori) [inline app/pkg/slice/string_test.go:24]: input の s が空スライスのパターン欲しいなと思った

File: app/internal/domain/service/slack_reaction_users.go [hunk 4]
```diff
@@ -43,13 +43,33 @@ func NewSlackReactionUsersService(factory repository2.Factory) *slackReactionUse
...
+const (
+	// ChunkSizeOfChunkedListUserEmail chunk size of calling slackRepository.ListUsersEmail
+	ChunkSizeOfChunkedListUserEmail = 20
+)
+
+// chunkedListUsersEmail splits userID array into chunks,
+//　and calls slackRepository.ListUsersEmail for each chunk.
+func (s *slackReactionUsersService) chunkedListUsersEmail(ctx context.Context, userIDs []string) ([]*model.SlackUserEmail, error) {
+	chunkedUserIDsList := slice.SplitStringSliceInChunks(userIDs, ChunkSizeOfChunkedListUserEmail)
+	var slackUserEmails []*model.SlackUserEmail
```
Comment (hotakahattori) [inline app/internal/domain/service/slack_reaction_users.go:55]:
```suggestion
	slackUserEmails := make([]*model.SlackUserEmail, 0, len(chunkedUserIDsList))
```

のほうが良さそう

**Review by hotakahattori — COMMENTED** [review]
@sho0126hiro 

comment shimashita.

File: app/pkg/slice/string.go (hunk 2 above)
Comment (sho0126hiro) [reply app/pkg/slice/string.go:40]: 個人的に、↓のような気持ちだから、もっと良い方法がなければここの部分はskipで良いかなと思ってる
- sliceパッケージを商切り上げをしたい、という理由で`math`packageに依存したくないという気持ち
  - makeを使う理由は、主にメモリの効率化が理由（あとは可読性？）だと思っているけど、mathを使うと返って性能悪くなる説をちょっと感じた
- math packageを使わない場合は、以下の条件分岐でcapを決められるんだけど、ここまでこだわりたいかが、あまりよく分かってない（pkgを利用する人が、返り値の長さを知りたいという理由なら、一旦メソッドのコメントに書いておくという方法もありそう）
  - `chunkSize < len(s)`の時に、`len(s)+1/chunkSize`
  - `chunkSize >= len (s)`の時に`1`

File: app/pkg/slice/string.go (hunk 1 above)
Comment (sho0126hiro) [reply app/pkg/slice/string.go:33]: thx! I wrote it: [add comment when function-user sets chunkSize is less than 1](https://github.com/moneyforward/auriga/pull/15/commits/c5e17165075586281e3fb3c0104a9e090194fac1)

File: app/pkg/slice/string_test.go (hunk 3 above)
Comment (sho0126hiro) [reply app/pkg/slice/string_test.go:24]: Thank you~ I did it: [add test when s is empty](https://github.com/moneyforward/auriga/pull/15/commits/6bdf1505a66f717014bd2949959ef46a6be0427f)

**Comment by sho0126hiro** [comment]
@hotakahattori  コメントありがとう〜、一つだけ対応せずに返信しました :bow:

File: app/pkg/slice/string.go (hunk 2 above)
Comment (hotakahattori) [reply app/pkg/slice/string.go:41]: ん、
cap＝(len(s)/chuknSize)+1

じゃダメかな？

Comment (sho0126hiro) [reply app/pkg/slice/string.go:41]: それだと `len(s)/chunkSize`が整数nの時、期待するcapの値はnだけど、上の式だとn+1になってしまいそう

Comment (hotakahattori) [reply app/pkg/slice/string.go:41]: lenではなくてcapだから、+1になるくらいは別に良いのでは？と思った(嫌だったら余りを見て、0じゃなかったら+1するとか？)

Comment (sho0126hiro) [reply app/pkg/slice/string.go:41]: それは思ったんだけど、そうなると、capを指定するべきなのかがよく分からなかった。性能的良さそう、という感じ?（やらないよりやった方が良い、という感じだったらやります〜）

Comment (sho0126hiro) [reply app/pkg/slice/string.go:42]: やっときました→[fix impl and add test](https://github.com/moneyforward/auriga/pull/15/commits/115a104f0ae46de9b661879c669a200bae1b4c31)

Comment (hotakahattori) [reply app/pkg/slice/string.go:42]: > 性能的に良さそう、という感じ?

そう

スライスに値を append していくと、スライス長が cap を超えたタイミングでスライスのメモリがアロケートされる (スライスをただ変数宣言しただけだと capは 0)
→ アロケートが頻繁に走ると、パフォーマンス悪い
→ cap が予め分かっているなら、予め指定しておいた方がパフォーマンス良い

これとか参考になりそう

https://qiita.com/imoty/items/bb18fb50d526474d2d10

Comment (sho0126hiro) [reply app/pkg/slice/string.go:42]: なるほど、1つ超える分には全然問題ないのか、ありがとう 👍
